src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,688 @@
+                        {
+                            source = GenerateForCollection(typeMetadata);
+
+                            GenerateTypeMetadata(typeMetadata.CollectionKeyTypeMetadata);
+                            GenerateTypeMetadata(typeMetadata.CollectionValueTypeMetadata);
+                        }
+                        break;
+                    case ClassType.Object:
+                        {
+                            source = GenerateForObject(typeMetadata);
+
+                            if (typeMetadata.PropertiesMetadata != null)
+                            {
+                                foreach (PropertyMetadata metadata in typeMetadata.PropertiesMetadata)
+                                {
+                                    GenerateTypeMetadata(metadata.TypeMetadata);
+                                }
+                            }
+                        }
+                        break;
+                    case ClassType.TypeUnsupportedBySourceGen:
+                        {
+                            _executionContext.ReportDiagnostic(
+                                Diagnostic.Create(TypeNotSupported, Location.None, new string[] { typeMetadata.CompilableName }));
+                            return;
+                        }
+                    default:
+                        {
+                            throw new InvalidOperationException();
+                        }
+                }
+
+                try
+                {
+                    _executionContext.AddSource($"{typeMetadata.FriendlyName}.cs", SourceText.From(source, Encoding.UTF8));
+                }
+                catch (ArgumentException)
+                {
+                    _executionContext.ReportDiagnostic(Diagnostic.Create(DuplicateTypeName, Location.None, new string[] { typeMetadata.FriendlyName }));
+                }
+            }
+
+            private string GenerateForTypeWithKnownConverter(TypeMetadata typeMetadata)
+            {
+                string typeCompilableName = typeMetadata.CompilableName;
+                string typeFriendlyName = typeMetadata.FriendlyName;
+
+                string metadataInitSource = $@"_{typeFriendlyName} = {JsonMetadataServicesClassName}.{GetCreateValueInfoMethodRef(typeCompilableName)}({OptionsInstanceVariableName}, {JsonMetadataServicesClassName}.{typeFriendlyName}Converter);";
+
+                return GenerateForType(typeMetadata, metadataInitSource);
+            }
+
+            private string GenerateForTypeWithUnknownConverter(TypeMetadata typeMetadata)
+            {
+                string typeCompilableName = typeMetadata.CompilableName;
+                string typeFriendlyName = typeMetadata.FriendlyName;
+
+                StringBuilder sb = new();
+
+                string metadataInitSource = $@"JsonConverter converter = {typeMetadata.ConverterInstantiationLogic};
+                    // TODO: consider moving this verification source to common helper.
```

Comment:
Do we really want to generate this TODO in the user's code?

Reply:
https://github.com/dotnet/runtime/pull/52147#discussion_r625398786